modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/IndexManager.java

```diff
@@ -281,6 +303,66 @@ boolean isIndexStale(ShpFileType indexType) {
     }
 
     /**
+     * Queries the spatial index for features available in the specified bbox, 
+     * applying the BBOX filter if specifided.
+     * 
+     * @param bbox
+     */
+    protected CloseableIterator<Data> querySpatialIndex(Envelope bbox, boolean applyEnvelopeFilter) throws DataSourceException, TreeException, IOException {
```

Comment:
I really hope I'm reading this wrong, but it seems that now in order to access a spatially indexed shapefile we have to scroll over it once to read the headers and build the Data list, and then go back at the beginning to read the actual geometries? Have you considered shapefiles that are not on the local disk, not fitting the file system cache, or coming from the classpath?